File: Homework/Objective-C/HW24/Superman/HW24/HW24/ViewController.m
//
//  ViewController.m
//  HW24
//
//  Created by Руслан on 22.10.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@property (assign, nonatomic) BOOL swiped;

@property (weak, nonatomic) IBOutlet DrawingView *mainDrawingView;
@property (weak, nonatomic) IBOutlet DrawingView *tempDrawingView;
@property (weak, nonatomic) IBOutlet DrawingView *viewBrush;

@property (weak, nonatomic) IBOutlet UILabel *redColorValue;
@property (weak, nonatomic) IBOutlet UILabel *greenColorValue;
@property (weak, nonatomic) IBOutlet UILabel *blueColorValue;
@property (weak, nonatomic) IBOutlet UILabel *opacityValue;
@property (weak, nonatomic) IBOutlet UILabel *brushValue;

@property (weak, nonatomic) IBOutlet UISlider *redColorSlider;
@property (weak, nonatomic) IBOutlet UISlider *greenColorSlider;
@property (weak, nonatomic) IBOutlet UISlider *blueColorSlider;
@property (weak, nonatomic) IBOutlet UISlider *opacitySlider;
@property (weak, nonatomic) IBOutlet UISlider *brushSizeSlider;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.mainDrawingView.delegate = self;
    self.tempDrawingView.delegate = self;
    self.viewBrush.delegate = self;
    
    [self refreshValues];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Helpful functions

- (void) touchEndedOrCancelled {
    
    if (!self.swiped) {
        [self.tempDrawingView drawLineFrom:self.lastPoint to:self.lastPoint];
        
    }
    [self.tempDrawingView mergeTempView:self.tempDrawingView intoMainView:self.mainDrawingView];
}

#pragma mark - Touches

- (void) touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    self.swiped = NO;
    
    UITouch *touch = [touches anyObject];
    self.lastPoint = [touch locationInView:self.tempDrawingView];
}

- (void) touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    
    self.swiped = YES;
    
    UITouch *touch = [touches anyObject];
    CGPoint currentPoint = [touch locationInView:self.tempDrawingView];
    
    [self.tempDrawingView drawLineFrom:self.lastPoint to:currentPoint];
    self.lastPoint = currentPoint;
}

- (void) touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self touchEndedOrCancelled];
}



#pragma mark - Actions

- (IBAction)changeValueSlider:(id)sender {
    [self refreshValues];
}

- (IBAction)erase:(id)sender {
    
    self.redColorSlider.value = 1.f;
    self.greenColorSlider.value = 1.f;
    self.blueColorSlider.value = 1.f;
    self.opacitySlider.value = 1.f;
    
    [self refreshValues];
}

- (IBAction)reset:(id)sender {
    self.mainDrawingView.image = nil;
}

#pragma mark - Refresh values

- (void) refreshValues {
    
    self.redColor = self.redColorSlider.value;
    self.redColorValue.text = [NSString stringWithFormat:@"%d", (int)(self.redColor*255)];
    
    self.greenColor = self.greenColorSlider.value;
    self.greenColorValue.text = [NSString stringWithFormat:@"%d", (int)(self.greenColor*255)];
    
    self.blueColor = self.blueColorSlider.value;
    self.blueColorValue.text = [NSString stringWithFormat:@"%d", (int)(self.blueColor*255)];
    
    self.opacity = self.opacitySlider.value;
    self.opacityValue.text = [NSString stringWithFormat:@"%1.1f", self.opacity];
    
    self.lineWidth = self.brushSizeSlider.value*100;
    self.brushValue.text = [NSString stringWithFormat:@"%1.1f", self.lineWidth];
    
    [self.viewBrush drawPreview];
}

@end

File: Homework/Swift/HW24/Superman/Superman/ViewController.swift
//
//  ViewController.swift
//  Superman
//
//  Created by Руслан on 23.10.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

import UIKit

class ViewController: UIViewController, DrawingViewDelegate {
    
    var swiped     =  false
 
// MARK: - protocole's properties
    
    var lastPoint  =  CGPoint()
    var redColor   =  CGFloat()
    var greenColor =  CGFloat()
    var blueColor  =  CGFloat()
    var opacity    =  CGFloat()
    var lineWidth  =  CGFloat()

// MARK: - IBOutlet properties
    
    @IBOutlet weak var mainDrawingView: DrawingView!
    @IBOutlet weak var tempDrawingView: DrawingView!
    @IBOutlet weak var viewBrush: DrawingView!
    
    @IBOutlet weak var redColorValue: UILabel!
    @IBOutlet weak var greenColorValue: UILabel!
    @IBOutlet weak var blueColorValue: UILabel!
    @IBOutlet weak var opacityValue: UILabel!
    @IBOutlet weak var brushValue: UILabel!
    
    @IBOutlet weak var redColorSlider: UISlider!
    @IBOutlet weak var greenColorSlider: UISlider!
    @IBOutlet weak var blueColorSlider: UISlider!
    @IBOutlet weak var opacitySlider: UISlider!
    @IBOutlet weak var brushSizeSlider: UISlider!
    
// MARK: - ViewController default functions
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        mainDrawingView.delegate = self
        tempDrawingView.delegate = self
        viewBrush.delegate = self
        
        refreshValues()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

// MARK: - Helpful functions
    
    func touchEndedOrCancelled() {
        
        if !swiped {
            tempDrawingView.drawLine(from: lastPoint, to: lastPoint)
        }
        tempDrawingView.merge(tempView: tempDrawingView, intoMainView: mainDrawingView)
    }
    
// MARK: - Touches
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        swiped = false
        
        let touch = touches.first!
        lastPoint = touch.location(in: tempDrawingView)
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        
        swiped = true
        
        let touch = touches.first!
        let currentPoint = touch.location(in: tempDrawingView)
        
        tempDrawingView.drawLine(from: lastPoint, to: currentPoint)
        lastPoint = currentPoint
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        touchEndedOrCancelled()
    }
    
    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        touchEndedOrCancelled()
    }
    
// MARK: - Actions
    
    @IBAction func changeValueSlider(_ sender: AnyObject) {
        refreshValues()
    }
    
    @IBAction func erase(_ sender: AnyObject) {
        redColorSlider.value = 1
        greenColorSlider.value = 1
        blueColorSlider.value = 1
        opacitySlider.value = 1
        
        refreshValues()
    }
    
    @IBAction func reset(_ sender: AnyObject) {
        mainDrawingView.image = nil
    }
    
// MARK: - Refresh values
    
    func refreshValues() {
        
        redColor = CGFloat(redColorSlider.value)
        redColorValue.text = "\((Int)(redColor * 255))"
        
        greenColor = CGFloat(greenColorSlider.value)
        greenColorValue.text = "\((Int)(greenColor * 255))"
        
        blueColor = CGFloat(blueColorSlider.value)
        blueColorValue.text = "\((Int)(blueColor * 255))"
        
        opacity = CGFloat(opacitySlider.value)
        opacityValue.text = String.init(format: "%1.1f", opacity)
        
        lineWidth = CGFloat(brushSizeSlider.value * 100)
        brushValue.text = String.init(format: "%1.1f", lineWidth)
        
        viewBrush.drawPreview()
    }
    
}














